src/libappimage/desktop_integration/integrator/ResourcesExtractor.cpp

```diff
@@ -0,0 +1,91 @@
+// system
+#include <map>
+
+// local
+#include <appimage/core/PayloadEntryType.h>
+#include "ResourcesExtractor.h"
+
+namespace appimage {
+    namespace desktop_integration {
+        namespace integrator {
+            ResourcesExtractor::ResourcesExtractor(core::AppImage appImage) : appImage(appImage) {}
+
+            DesktopIntegrationResources ResourcesExtractor::extract() {
+                DesktopIntegrationResources resources = {};
+                for (auto fileItr = appImage.files(); fileItr != fileItr.end(); ++fileItr) {
+                    if (fileItr.type() != core::PayloadEntryType::REGULAR && fileItr.type() != core::PayloadEntryType::LINK)
+                        continue;
+
+                    const auto& filePath = *fileItr;
+
+                    if (extractDesktopFile && isMainDesktopFile(filePath)) {
+                        resources.desktopEntryPath = filePath;
+                        resources.desktopEntryData = std::vector<char>(std::istreambuf_iterator<char>(fileItr.read()),
```

Comment:
We can save the  `std::vector<char>` and use a list initialization constructor. But we still need the `istreambuf_iterator` to read from an `stream` to a `vector`.

The descrived changes where applied.

Reply:
What I mean is, maybe `read` should just perform the reading, and return the vector. We could add another method to allow access to the stream buf.

Reply:
An `std::istream` is used to don't force the user to allocate the whole file in memory if not required.  In this case, we do require to store the whole file but if we are going to create a helper for it let's do it in the `ResourcesExtractor` class. `PaylodIterator` is already quite complex to add more stuff.

Reply:
I disagree. The method belongs to the iterator. Or some sort of "underlying" object that I'm still missing (i.e., we still return a string in `operator*`, but could/should probably return some sort of object that allows for quick access to the file (position, etc., something archive specific (not type, but archive)).

It might generally be better to abstract filesystem type from AppImage type. Right now, they're too tightly coupled.

Reply:
I guess that we better continue this on #64